### PR TITLE
Use connctionhelper to connect to remote Docker host.

### LINKED ISSUE #580: Add support for SSH-based connection to Docker daemon
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

The `docker` CLI supports SSH-based connections, which greatly simplifies the use of remote systems for performing Docker operations, such as building images (very useful, for example, for building AMD64 images from an ARM64-based local system). However, the Docker provider does not seem to support SSH-based connections; it returns an error when attempting to use a `DOCKER_HOST` environment variable set to an SSH-based connection.

### Affected area/feature

Docker provider

### LINKED ISSUE #662: Can't connect to remote Docker via SSH
### What happened?

Hi. First of all, thanks for making Pulumi! I think it's pretty amazing tool.

I'm trying to deploy one of my applications to DigitalOcean server using Pulumi + Docker via SSH connection. While from Docker CLI this seem to work and pretty simple to set up, pulumi fails to build an image on remote server. From the look of the error I'm getting, it seems like after ssh connection is established, pulumi (or docker?) completely ignoring the fact I'm working via SSH, and proceeds building image via http for some reason. You can find more information by the link below. This happens at least when I'm using pulumi YAML runtime.

### Expected Behavior

Docker connection to remote server is succeeded (which is seem to work), my image is sent and built on that server, and container is up and running (which doesn't work).

### Steps to reproduce

You can find all steps in this repository: https://github.com/octet-stream/pulumi-docker-ssh-issue

It also includes a simple example TCP server and a Dockerfile to build an image for that server.

### Output of `pulumi about`

CLI          
Version      3.72.2
Go Version   go1.20.5
Go Compiler  gc

Plugins
NAME    VERSION
docker  unknown
yaml    unknown

Host     
OS       darwin
Version  13.4
Arch     arm64

This project is written in yaml

Current Stack: octet-stream/pulumi-docker-ssh-issue/dev

TYPE                     URN
pulumi:pulumi:Stack      urn:pulumi:dev::pulumi-docker-ssh-issue::pulumi:pulumi:Stack::pulumi-docker-ssh-issue-dev
pulumi:providers:docker  urn:pulumi:dev::pulumi-docker-ssh-issue::pulumi:providers:docker::default
pulumi:providers:docker  urn:pulumi:dev::pulumi-docker-ssh-issue::pulumi:providers:docker::docker-remote

Found no pending operations associated with dev

Backend        
Name           pulumi.com
URL            https://app.pulumi.com/octet-stream
User           octet-stream
Organizations  octet-stream

No dependencies found

### Additional context

Along with the `pulumi about` information: I'm using docker `v24.0.2` on the client and `19.03.14` on the server.

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

### PR DESCRIPTION
**Pull request by guineveresaenger**
Fixes #662 
Fixes #580

### CODE REVIEW

File: provider/go.mod [hunk 1]
```diff
@@ -9,15 +9,15 @@ require (
...
 	github.com/moby/buildkit v0.11.4
 	github.com/moby/moby v23.0.1+incompatible
-	github.com/moby/patternmatcher v0.5.0
 	github.com/pkg/errors v0.9.1
-	github.com/pulumi/pulumi-terraform-bridge/v3 v3.47.3
-	github.com/pulumi/pulumi/pkg/v3 v3.66.0
-	github.com/pulumi/pulumi/sdk/v3 v3.66.0
-	github.com/stretchr/testify v1.8.2
+	github.com/pulumi/pulumi-terraform-bridge/v3 v3.50.1
+	github.com/pulumi/pulumi/pkg/v3 v3.71.0
```
Comment (t0yv0) [inline provider/go.mod:14]: Would be best if the updates were a separate (automated) PR.

File: provider/image.go [hunk 2]
```diff
@@ -813,13 +813,26 @@ func configureDockerClient(configs map[string]string) (*client.Client, error) {
 
 	// No TLS certificate material provided, create an http client
 	if host != "" {
+		// first, check for ssh host
+		sshopts := []string{}
+		helper, err := connhelper.GetConnectionHelperWithSSHOpts(host, sshopts)
+		if err != nil {
+			return nil, err
+		}
+		if helper != nil {
```
Comment (t0yv0) [inline provider/image.go:822]: I don't understand this bit here, is there ever the case where helper==nil? Oh

```It returns nil without error when no helper is registered for the scheme.```

**Comment by t0yv0** [comment]
This needs tests. Thoughts on adding tests? I understand that standing up an SSH docker host is perhaps problematic for integration testing here? Can we consider factoring out the code to use unit tests, or is it possible to run ProgramTest with an SSH based docker host without actually standing it up, and observe some meaningful difference in before/after the change that we can assert on?

File: provider/go.mod (hunk 1 above)
Comment (guineveresaenger) [reply provider/go.mod:14]: I can run that first if you prefer

Comment (guineveresaenger) [reply provider/go.mod:14]: https://github.com/pulumi/pulumi-docker/pull/669

**Review by AaronFriel — DISMISSED** [review]

**Comment by AaronFriel** [comment]
Agree with @t0yv0 on wanting to test this:

>  Can we consider factoring out the code to use unit tests,

This seems like the lightest weight approach; can we create docker contexts in a unit test (without any running Docker host) and test that the helper uses the docker context selected?

File: provider/image.go (hunk 2 above)
Comment (guineveresaenger) [reply provider/image.go:822]: added explanation!

**Comment by guineveresaenger** [comment]
I added a few unit tests that verify ssh-based behavior.

It turns out that the SSH connection helper _only_ supports ssh-based scheme, and it _always_ returns `http://docker.example.com` as the client's daemon host.

I leverage this knowledge in the tests as checking for `http://docker.example.com` as the daemon host, and added a few default checks 🤞

**Review by t0yv0 — APPROVED** [review]

**Comment by necaris** [comment]
I hope this will make it into a 4.2.5 that will be released soon?

**Pull request merged**